KEP-4671: Update Workload API and Gang Scheduling KEP for beta in v1.37 (#6076)
kubernetes/enhancements: keps/sig-scheduling/4671-gang-scheduling/README.md

Reviewer: Shouldn't we get rid of all these 1.36 references and instead focus on current shape?

Author: Removed unnecessary sentences, PTAL

Reviewer 2: ```suggestion
#### Status Transition Rules
```

Reviewer 3: nit: better to mention that now each graduation needs to meet the criteria mentioned on KEP-5710 side too.

Author: Good point, added

Reviewer 4: I believe you'll need to update this, b/c you're dropping v1alpha2, right?

Reviewer 5: But it was introduced as part of this KEP.
So technically we added and removed - if you don't enable alpha APIs - it doesn't matter.
And if you do - you do that on your own risk anyway...

Author: Added a few sentences here describing what changed

Reviewer 4: Can you fix this typo here, should be `dependencies`

Reviewer 4: Can you update the `XX` numbers in this and followup section with actual values?

Author: IIUC these `XX/s` numbers mean two-digit values for the number of API calls per second. It's hard to provide some numbers here, as they depend on the exact configurations. Whatever concrete number we put here, it would be more or less random

Reviewer 4: Gotcha, I wasn't quite sure the meaning behind the `XX/s`, so thx for the explanation, that makes sense.

New core types called `Workload` and `PodGroup` are introduced to tell the kube-scheduler that a 
group of pods should be scheduled together and to define policy options related to gang scheduling. 
Pods may have an object reference in their spec to the `PodGroup` they belong to. The `Workload` 
and `PodGroup` objects are intended to evolve[^2] via future KEPs to support additional 
kube-scheduler improvements, such as topology-aware scheduling.

> In the API:

Reviewer 6: Can we update it to something like:
```
The API is structured to decouple policy from runtime grouping:
```
rather than the bare "In the API"? The reader who is unfamiliar with the history might benefit from it.

Author: Good idea, updated

## Proposal

> This KEP introduces both the `Workload` and `PodGroup` APIs in `scheduling.k8s.io/v1beta1`. The 1.36 revision, 

Reviewer 5: nit: I would remove the "/v1beta1" here - let's focus on the API group, as versions will change over time as it's visible here.

type GangSchedulingPolicy struct {
	// MinCount is the minimum number of pods that must be schedulable or scheduled
	// at the same time for the scheduler to admit the entire group.
	// It must be a positive integer. This field is mutable to support workload scaling.

>   //

Reviewer: NIT: Indent

>   // Note that the scheduler operates on an eventually consistent model. Updates

Reviewer 2: Nitpick - indentation.
```suggestion
	//
	// Note that the scheduler operates on an eventually consistent model. Updates
```

    // Conditions represent the latest observations of the PodGroup's state.
    //
    // Known condition types:

>     // - "PodGroupInitiallyScheduled": Indicates whether the scheduling requirement has been satisfied.

Reviewer 7: Frankly I'm not sure if adding "initially" make it any cleaner. I feel like the original name is good enough to reflect that it holds the initial status.

I'd suggest to keep it consistent with Pod status as I don't see a clear justification for this change.

Reviewer 5: Pod - once scheduled remains scheduled.
This is not true for PodGroup, where individual pods may come and go.

PodGroupScheduled suggests that the group remains scheduled, and it may not really be the case.

PodGroupInitiallyScheduled may not be the best name and I'm open to hear other suggestions, but I'm pretty sure it better reflects what it is that PodGroupScheduled.

Reviewer 7: Ack.

### PodGroup Status Lifecycle

The `PodGroup.Status` is managed by kube-scheduler to reflect the scheduling status. We introduce
a `Conditions` field with the `PodGroupInitiallyScheduled` condition type.

`PodGroup` status mirrors `Pod` status semantics rather than defining PodGroup-specific reasons:
- If pods are unschedulable (i.e., timeout, resources, affinity, etc.), the scheduler updates the 
`PodGroupInitiallyScheduled` condition to `False` and sets the reason fields accordingly.
- If pods are scheduled, the scheduler updates the `PodGroupInitiallyScheduled` condition to `True` after the group 

> got accepted by the Permit phase.

Reviewer: We should get rid of / replace those Permit references if we're planning to retire its usage with PlacementFeasible.

Author: In this specific case, the mention of Permit was incorrect. Fixed the wording.

However, currently we are not planning to get rid of Permit. It still works as a barrier before the pods are send to binding, to reverify the constraints.

For basic scheduling policy, when the pod related to the `PodGroup` gets scheduled, the scheduler updates 
the `PodGroupScheduled` condition to `True`.

#### Status Transition Rules

Once a `PodGroup` transitions to `PodGroupScheduled=True`, it is treated as a terminal scheduling 
state and does not revert to `False`. Specifically, once the group's scheduling constraint (`minCount`) 
has been satisfied, subsequent failed scheduling cycles for additional pods beyond `minCount` do not 
regress the condition. On same-status transitions (e.g., `True` → `True`), the condition message may 
be updated, but `LastTransitionTime` remains unchanged.

> This is a deliberate simplification. In practice, scheduled pods may later be evicted

Reviewer: I agree we shouldn't track runtime status here, but calling it a "deliberate simplification" leaves the actual semantics ambiguous. Here is how I see it:

1. First, a status condition can reflect either scheduling decisions or runtime state. For a single Pod, these effectively mean more or less the same thing because a Pod is scheduled only once. For a PodGroup, this is not true. A PodGroup can be evaluated multiple times as new pods appear, triggering new scheduling cycles.

2. Second, the scheduler currently lacks the mechanisms to track runtime state (e.g., evictions, node failures). If we ever want the scheduler to handle rescheduling and broader runtime failure handling, it will be a paradigm shift requiring a separate KEP. For now, we must operate within the existing framework and restrict this status strictly to scheduling decisions.

3. However, even with these boundaries established, we cannot just leave it as is. The exact scheduling semantics remain unclear. We need to explicitly choose and document one of two approaches:

   1. Latched Status: The condition turns True once minCount is satisfied and never reverts. Crucially, if some pods are later deleted and new pods are added, the scheduler runs a new cycle. Even if this new cycle fails to satisfy minCount, the status still remains True.

   2. Latest Cycle Status: The condition always reflects the outcome of the most recent scheduling cycle. If new pods are added, triggering a new cycle, the status updates to reflect whether the PodGroup is currently schedulable (e.g., it reverts to False if the current state no longer satisfies minCount).

Regardless of which option we choose, we must explicitly document these exact semantics instead of waving them off as a beta simplification. Depending on the choice, we also need to ensure the implementation actually matches it (e.g. if we go with 2). 

Personally, I think I like 2 better given it provides up2date scheduling decision / status.

Reviewer 5: I was going back and forth with writing something similar and finally decided that we can live with (1). But reading your comment now, I think that it's still probably OK *if* we would adjust the name of the condition.
`PodGroupScheduled` suggests the current state and that is not really what it is reflecting.
So maybe we should rename it to something different (PodGroupScheduledOnce? - I don't like it but I don't have a better name).

Latest Cycle Status - I don't like this semantic, because it reflects last "scheduling attempt" not really what is happening with the group. The group can technically be fully scheduled (satisfying all constraints including minCount) but e.g. the additional pod (above minCount) couldn't scheduled.
If we want something like that - the name should clearly reflect that it's not about current state, but rather the status of the last scheduling cycle.

Author: > The group can technically be fully scheduled (satisfying all constraints including minCount) but e.g. the additional pod (above minCount) couldn't scheduled.
> If we want something like that - the name should clearly reflect that it's not about current state, but rather the status of the last scheduling cycle.

+1, Also, even if the minCount is no longer met, the PodGroup has scheduled pods that may run, and in case of TAS, the selected placement which new pods will follow. So, I think it's beneficial to keep the current condition.

If we need the "latest cycle status", we could create a new condition to capture it pretty easily.

> So maybe we should rename it to something different (PodGroupScheduledOnce? - I don't like it but I don't have a better name).

`PodGroupInitiallyScheduled`?

Reviewer: > Latest Cycle Status - I don't like this semantic, because it reflects last "scheduling attempt" not really what is happening with the group. The group can technically be fully scheduled (satisfying all constraints including minCount) but e.g. the additional pod (above minCount) couldn't scheduled.

I think my previous point might not have been entirely clear. In this exact scenario, in option 2., the group should have the Scheduled status because the user intent (minCount) is satisfied. Individual pods above the minCount might remain Unscheduled, but the group as a whole meets the user's specified requirements. 

To me, this is the "relatively cleanest" approach and I would argue that Option 2 actually reflects "what is happening with the group" better than Option 1. As noted above, if minCount is met, the user intent is satisfied and the group is considered scheduled, even if additional pods are not. With Option 1, the user loses visibility after the first successful scheduling attempt. If a disruption occurs, pods terminate, and the scheduler subsequently fails to place the recreated pods (dropping the feasible count below minCount), the status remains Scheduled. This is misleading. Option 2 updates the status based on whether minCount is satisfied in the most recent cycle. This provides an up-to-date answer to whether the user intent is currently met by the scheduler. And obviously this also requires thorough documentation regarding its main limitation. Because it is strictly a scheduling status, if pods are disrupted and not recreated (meaning no new scheduling cycle is triggered), the status will not update. This can also be confusing.

If we decide to go with latching the first successful state, I agree we must rename the condition and document its semantics thoroughly. @macsko's proposal (PodGroupInitiallyScheduled) seems fine. It accurately reflects that we are recording the first successful scheduling attempt (even if preceded by multiple failed attempts before) and ignoring subsequent state changes.

Reviewer 5: OK - so it's not really about "Last Cycle State", but whether the PodGroup satisfies the conditions.
I agree this will be useful, but the primary question is whether it's really kube-scheduler who should be setting that. The "healthiness" of PodGroup can break as a consequence of e.g. disruption of a pod not even triggered by scheduler and if either there will not be recreated pod it will never be updated.

To be clear - I think we actually need both (Maciek above pointed more good arguments why the current condition is beneficial - e.g. locked in TAS placement).
The primary question is - what is a beta blocker and what can be added later.

Personally, I'm leaning towards say that:
1) we keep the existing condition and only that is the beta blocker - we just rename it - I'm fine with `PodGroupInitiallyScheduled`
2) we postpone the "PodGroup status" for now and will address it in a follow-up

Author: Initially, I thought it wouldn't be simple to handle the "latest cycle status" in the code, but I realized that with the new PlacementFeasible extension point it will be fairly easy. So, apart from API changes (description?), it should be quick to implement the support for both statuses.

I see the benefits of both options, so I think the primary question is even whether we have the capacity to implement the "latest cycle status" this release or not.

> I agree this will be useful, but the primary question is whether it's really kube-scheduler who should be setting that. The "healthiness" of PodGroup can break as a consequence of e.g. disruption of a pod not even triggered by scheduler and if either there will not be recreated pod it will never be updated.

I think we are mixing now different concepts. In my opinion, live tracking the state (maybe including the current pod counts, I remember such idea being raised some time ago) of the pod group shouldn't be done by kube-scheduler, but rather a different component. The scheduler, however, can display its current decision in the status, which is also important.

or impacted by node failures, but the `PodGroup` status will not track
these post-scheduling disruptions. While a more robust status lifecycle capable of reflecting
post-scheduling state changes is a desirable future enhancement, we do not consider it
a blocker for beta graduation. Handling such behavior is tied to rescheduling 

> mechanisms, which remain out of scope for this KEP.

Reviewer 5: I would remove the last sentence - this is not necessarily tied to rescheduling. It may be used to trigger other actions by related, but different components.

Also, currently kube-scheduler, why it tracks pod terminations/removals to ensure its cache reflects the available resources, it doesn't take any external actions based on those.
So arguably, we should consider having a separate (at least logically) component for that.

look up the `PodGroup` object. The scheduler requires `PodGroup` object to exist before scheduling 
pods that reference them.

### Pod Group minCount Mutability

> 

Reviewer: I would clarify that minCount is a "scheduling constraint" and not a "runtime constraint". This means if there is a scheduled PodGroup with minCount=10 and 10 pods, and the minCount is increased to 20 without creating new pods, those 10 pods will remain running even though the condition is no longer met. Consequently, given how we report PodGroup status, it will also remain "Scheduled".

Author: We have "`minCount` is enforced strictly during scheduling, not at runtime. This means modifications to the field will not affect already-scheduled pods, applying only to those evaluated in future scheduling cycles."

To support workload scaling (e.g. for elastic Jobs), the `minCount` field in both `PodGroup`
and `PodGroupTemplate` is proposed to be mutable in v1.37. Specifically:

* Modifying a `PodGroupTemplate` will not affect existing `PodGroup` instances.
  Changes should apply exclusively to new instances created from the updated template by the controllers.

* Because the scheduler operates on an eventually consistent view of the cluster state,

>   updates to `minCount` may not be immediately visible. If a modification occurs while the scheduler

Reviewer 7: ```suggestion
* In case of `PodGroup` instances, updates to `minCount` may not be immediately visible,
as the scheduler operates on an eventually consistent view of the cluster state. If a modification occurs while the scheduler
```

Author: Changed

  is actively processing that specific `PodGroup` in a scheduling cycle, the previous value
  may still be evaluated for that cycle. The new value will take effect in the subsequent cycle.

* `minCount` is enforced strictly during scheduling, not at runtime. This means modifications to the field
  will not affect already-scheduled pods, applying only to those evaluated in future scheduling cycles.

While broader PodGroup spec mutability, such as modifying the number of `PodGroupTemplates` in a `Workload`,
may be desirable, we are strictly scoping API mutability in v1.37 to the `minCount` field.
Further relaxation of validation rules will be considered in the future if driven by strong use cases.

> ### Workload Controllers Integration

Reviewer 6: We should add link to the KEP (#5547). And it would be good to add feedback from these projects about the integration or list issues.

Author: Added link to the KEP. Regarding adding the feedback, I don't think we need that level of details, listing the exact feedback. The confirmation that we go in a good direction and don't miss anything should be enough, and further discussion and design should be done on the controllers side. We don't want to keep such section being constantly updated based on the controllers integration stage.

look up the `PodGroup` object. The scheduler requires `PodGroup` object to exist before scheduling 
pods that reference them.

> ### Pod Group minCount Mutability

Reviewer 2: Let's say we have a gang PodGroup which was scheduled and its Pods are already running for some time.

Now, let's assume the minCount value for that group got bumped up, new Pods appear but the scheduler is not able to fit those even with preemption.

What is going to happen with the already running Pods? Are they going to be evicted by the scheduler or will they continue to run as they are?

Author: Good point, added a sentence about this to the field documentation and to this KEP section. The already-scheduled pods should remain intact in such case

The single scheduling cycle, together with blocking resources using nomination,
will address requirement (3).

> We will also introduce [Delayed Preemption](#delayed-preemption).

Reviewer 7: Maybe we could call the section "### Scheduler Changes release history" and describe briefly 1.37 scope:
- drop delayed preemption in favor of Workload Aware Preemption
- introduction of minCount mutability
- most of all, promotion to beta
- new extension point
- feature gates merge
- etc.

Author: This KEP is already quite complicated and contains multiple sections written based on the specific releases' states. I would like to avoid having further sections like this that will be outdated soon anyway.

Reviewer 7: This section already contains a history of changes for 1.35 and 1.36, so maybe we should either drop it as well?

Author: I adjusted this section now to minimize references to specific versions. I know that we could get rid of more paragraphs, but it could be done in a cleanup later.

     Workload Scheduling Cycle to verify that the released resources make the placement feasible.
   * If `minCount` cannot be met (even after calculating potential
     preemptions), the scheduler considers the `PodGroup` unschedulable. Standard backoff
     logic applies (see *Failure Handling*), and `PodGroup` is returned to
     the scheduling queue.

#### Queuing and Ordering

> Workload-aware preemption ([KEP-5710](https://github.com/kubernetes/enhancements/pull/5711))

Reviewer 4: It's best to link to the tracking KEP issue https://github.com/kubernetes/enhancements/issues/5710 rather than actual PRs 😉

     Workload Scheduling Cycle to verify that the released resources make the placement feasible.
   * If `minCount` cannot be met (even after calculating potential
     preemptions), the scheduler considers the `PodGroup` unschedulable. Standard backoff
     logic applies (see *Failure Handling*), and `PodGroup` is returned to
     the scheduling queue.

#### Queuing and Ordering

Workload-aware preemption ([KEP-5710](https://github.com/kubernetes/enhancements/pull/5711))
will introduce a specific scheduling priority for a workload.

> Having that in mind, the queueing mechanism should support the workload's scheduling priority.

Reviewer 5: PodGroup's scheduling priority

[this is what scheduler cares about]

Author: I rewrote that paragraph completely based on a discussion from different thread

>    Based on the cycle's outcome:

Reviewer 7: I think we have just added a case in which PodGroup goes directly into ActiveQ after scheduling attempt at least in the following cases:
- preemption above minCount
- new pods that appeared during scheduling

Author: Good point. Added sentences about both

     using the nodes selected during the Workload Scheduling Cycle.

     The `WaitOnPermit` gate is retained to ensure that the `minCount` pods are successfully
     admitted before binding occurs. Additionally, the `minCount` check can consider
     the number of pods that have passed the Workload Scheduling Cycle to ensure
     that Pods do not wait unnecessarily if some have been rejected while new pods
     have been added to the cluster.

     In this successful case, preemption will not be attempted even if some pods remain unschedulable
     This means that as long as the scheduling constraints (`minCount`) are met, the schedulable pods
     will move to binding. Any subsequent preemption for unschedulable pods will be handled in the next
     [Workload Scheduling Cycle](#workload-scheduling-cycle) if still needed. Triggering both binding
     and preemption in the same cycle would be ambiguous, and such precedence wouldn't be clear
     from the scheduler's perspective. Alternatively, always attempting preemption to free up space,
     even for schedulable groups, would be unnecessarily disruptive and delay the startup of the group. 
     

>      Unschedulable pods will be requeued using their old timestamp, meaning the subsequent scheduling

Reviewer: Let's make it clear what condition will actuallu trigger preemption for them. 

According what we have in the text, the preemption will not be triggered in the next cycle. 

Let's assume we have a PG with 15 pods and `MinCount=10`. If 12 pods succeed in the first pass and 3 fail, the flow looks like this:

1. `schedulableCount >= 10`: We bind the 12 pods, do not trigger preemption, and finish the cycle.
2. The 3 remaining pods are requeued, so the PG enters the scheduling cycle again.
3. We still have `schedulableCount >= 10`, so according to the condition preemption is not triggered

Reviewer 5: Good callout.
The condition should actually be:
1) not every pod has been scheduled
and
2) it is NOT an initial scheduling of the PG

The alternative for (2) might be "we didn't find a place for any pods", but if the PG is already scheduled, I think we should prioritize finding the place for as many pods as possible over quickly adding a subset of pods (and soon after adding more).
The reasoning is that adding more pods to a gang generally has some overhead anyway (e.g. requires pod restart) so let's avoid triggering this twice.

Author: I'll adjust this section to capture this scenario correctly.

> The alternative for (2) might be "we didn't find a place for any pods", but if the PG is already scheduled, I think we should prioritize finding the place for as many pods as possible over quickly adding a subset of pods (and soon after adding more).
> The reasoning is that adding more pods to a gang generally has some overhead anyway (e.g. requires pod restart) so let's avoid triggering this twice.

I don't understand that you meant here. Could you clarify how  "it is NOT an initial scheduling of the PG" is better than "we didn't find a place for any pods"? Is is that because we already met the contract (scheduled minCount), so we can try the preemption, even when some new pods are schedulable?

How would that play with the case where the minCount is no longer met (some pods disappeared) and the scheduler has a decision to either schedule up to minCount (and try preemption in subsequent cycle for other pods) or preempt (without scheduling schedulable pods)? Also, what would we want to do with the basic policy?

Reviewer 5: > Is is that because we already met the contract (scheduled minCount), so we can try the preemption, even when some new pods are schedulable?

I tried to explain in my second paragraph, but let me try to clarify.
We need to take a step back and think what effectively happens at the application level when we change the size of the gang. In those cases we're generally talking about cases, where individual pods within a gang talk to other pods. So changing the size of a gang effectively requires "reconfiguration of the workload at runtime".
Different frameworks solve that differently, but this isn't instant (it may require in-place restarts, it may be solved by distributing some config and reinitializing etc.). But it's not instant.
From the perspective, having multiple cases where we add the next (set of) pod(s) soon after the other is simply not efficient (due to hit on the "goodput" of those workloads).

So once the workload is running, we should rather try to find place for as many remaining pods as possible and them shoot them into the workload in one batch.

> How would that play with the case where the minCount is no longer met (some pods disappeared) and the scheduler has a decision to either schedule up to minCount (and try preemption in subsequent cycle for other pods) or preempt (without scheduling schedulable pods)?

That's a great question. And clearly we can do better in this case.
But that's more of a corner case to me (which will happen of course, but shouldn't be super frequent).
So I would suggest ignoring that for now, getting some production experience with the first simple version of it and react based on users feedback.

> Also, what would we want to do with the basic policy?

I would bet on simplicity at the first step - let's do the same. And in the future we can iterate on that based on the feedback.

Author: > So once the workload is running, we should rather try to find place for as many remaining pods as possible and them shoot them into the workload in one batch.

Agree, trying preemption is such cases sounds like a reasonable approach. However, this might have a tricky outcome when pods from a single group are constantly coming and triggering additional preemptions. For example, we could have:
- Scheduler already scheduled 5 pods of pod group with minCount=5
- New pods from the pod group are coming
- 1 pod is observed, triggers the unsuccessful scheduling cycle and runs preemption
- 1 more pod is observed, while the previous pod is waiting for victims to terminate
- Scheduling cycle is triggered, 1 pod can be scheduled (as it freed up space via preemption), but another pod cannot. The scheduler tries yet another preemption and moves both pods to scheduling queue.
- 1 more pod is observed ...

Such scenario might significantly delay the binding of these pods.

Reviewer 5: Agree - but that sounds a bit like a maliciously constructed problematic scenario :)

I would ignore it for now - let's try to graduate what solves huge majority of problems and then iteratively improve it in follow-ups.

Author: Alright, if we say that we accept such "malicious" scenarios, I'm fine. Let me update the KEP, so the scheduler will follow a rule that if:

- not every pod has been scheduled AND
- at least one pod was already scheduled at the beginning of the scheduling cycle (i.e., it's not an initial attempt, but any subsequent one)

then we trigger the preemption, even if some pods may be schedulable without it.

Author: Added, PTAL

>      and preemption attempt should start immediately after the current cycle,

Reviewer: Running preemption in subsequent cycles effectively means running preemption for an already scheduled `PodGroup`. I think this becomes problematic for TAS for three reasons:

1. This is similar to what you pointed out in the CPG KEP discussion ([#6017](https://github.com/kubernetes/enhancements/pull/6017#discussion_r3304252339)), we choose a placement based on the current number of schedulable pods. In subsequent cycles, those placements may not fit extra pods. Conversely, running preemption from the beginning allows us to choose alternative placements that can accommodate the entire `PodGroup`.
2. The current design for the TAS preemption algorithm relies on re-generating placements. However, in this scenario, we cannot re-generate placements because they are already locked in for the currently running pods and cannot be changed.
3. To be clear, this issue exists today and would exist anyway even if we went with a single greedy cycle with preemption - this will happen anyway if pods are added to a `PodGroup` later. We likely need to accept this tradeoff and acknowledge that it won't work ideally in all cases, which is simply a property of TAS+WAP. This stems from the fact that pod-level preemption is much simpler because the preemptor itself is never already scheduled, which is no longer true for `PodGroups`.

@wojtek-t @Argh4k

Reviewer 5: Yes - I agree it will be suboptimal and may lead to unschedulability of those additional pods. But it doesn't break things.

I commented on that in https://github.com/kubernetes/enhancements/pull/6017#discussion_r3309809542 already

Author: Yes, TAS is an example of a behavior where trying the preemption, even with schedulable pods, would be beneficial. However, the motivation for not doing the preemption still holds - we don't want to delay the startup of schedulable pods. So, I agree that we have to accept this tradeoff and document it on the TAS + WAP level.

> 1. This is similar to what you pointed out in the CPG KEP discussion ([#6017](https://github.com/kubernetes/enhancements/pull/6017#discussion_r3304252339)), we choose a placement based on the current number of schedulable pods. In subsequent cycles, those placements may not fit extra pods. Conversely, running preemption from the beginning allows us to choose alternative placements that can accommodate the entire `PodGroup`.

It's different, because compared to the CPG case, preemption is much heavier and may cause significant delays, waiting for victims to terminate.

> 2\. The current design for the TAS preemption algorithm relies on re-generating placements. However, in this scenario, we cannot re-generate placements because they are already locked in for the currently running pods and cannot be changed.

In such case, TAS will generate only one placement, matching already-scheduled pods, so the preemption will consider victims only in this specific placement.

   It also utilizes the Opportunistic Batching feature where possible,
   reducing overall scheduling time.

   * If a pod fits, it is temporarily assumed and reserved on the selected node.

   * If a pod cannot fit, it is considered unscheduled for the current cycle. Preemption evaluations for individual pod
     is deferred until after the scheduler attempts to place all pods in the group, ultimately executing a unified step
     via [Workload-aware Preemption](#workload-aware-preemption). Meanwhile, the scheduling of subsequent pods continues
     as long as the `minCount` constraint remains satisfiable. The processing can also be optimized by rejecting
     all subsequent pods from the same homogeneous sub-group, as their failed scheduling outcome will be the same.

   The phase can effectively stop once `minCount` pods have a placement,
   though attempting to schedule the full group is preferred to maximize utilization.

4. The scheduler checks if the total number of schedulable pods meets the `minCount`.

   * If `schedulableCount >= minCount`, the cycle succeeds. Pods proceed directly to their binding cycles

>      using the nodes selected during the Workload Scheduling Cycle.

Reviewer 5: What we've been talking offline couple weeks ago was if we don't have to run the preemption here.

If we managed to schedule at least minCount pods, the group is technically schedulable.
But if there remained some unscheduled pods and we can actually make place for them via preempting other lower-priority workloads, we should actually still do that.

Now the remaining question is whether:
1) we immediately send the schedulable pods to binding
or
2) we actually trigger preemption to find space for the missing ones, because as part of that we may actually find a more optimal space for the whole group

The answer to that is not a clear tradeoff to me.
On one hand I would like to have better placement, but on the other hand preemption may take significant time and I don't want to unnecessary delay startup of this workload.
But I would personally lean towards not unnecessary delaying startup of those pods.

So I believe what should happen is:
1) if schedulable >= minCount the cycle succeeds and pods move to binding
2) if some pods remained unschedulable - trigger the preemption for them
  [now the question is whether it's the same cycle - which would be a bit weird, because the cycle technically already succeeded; or we somehow immediately trigger the next cycle to handle the remaining pods]

@macsko @sanposhiho @dom4ha @mm4tt

Reviewer 8: > because as part of that we may actually find a more optimal space for the whole group 

Is that relevant for the non TAS pod scheduling cycle? What would more optimal placement in a non TAS case mean? I think that actually this problem is more of a TAS problem, whether we should take placement that meets minCount or we should try to look for a placement that with preemption can give us placement that will fit all pods.

Reviewer 5: TAS is the clearest example, but the exact same thing you can achieve with pod affinity/anfi-affinity.

But even ignoring hard constraints, we can phrase this same problem in the context of scores - if we approximate score(PodGroup) as sum of scores of individual pods for non-TAS case, with preemption we may be able to find visibly better place.
Now - we don't do that if we can find a place without preemption, but given we didn't find it for all pods and we need to preempt something, we may actually find a better one (e.g. if we will preempt some large low-priority job).

Author: I think it depends on how we treat the minCount. If we say that the minCount is the main schedulability condition and later pods can be considered best-effort, then we can send schedulable pods to binding, move unschedulable pods back to the scheduling queue, and then trigger the preemption in a subsequent scheduling cycle, if it's still needed. I would prefer this option, as it would be the easiest to implement and, I think, pretty easy to understand by the users.

The problem would be when we say that we **always** want to find the best placement out of the current cluster state for **all** pods that we have. Obviously, it still might not be the best placement because we (will) optimize the number of nodes/placements to score, but the preemption could help us achieve something better than we have without it. In this case, I wouldn't like to send subset of pods to binding and trigger the preemption for the rest. It would be unintuitive in the code. However, if we run the preemption with all the pods, we may unnecessarily delay the startup of schedulable pods, because they would have to come back to the scheduling queue to wait for the victims to terminate.

Given that, I'm leaning towards executing the preemption in a subsequent cycle for unschedulable pods.

Reviewer 5: > Given that, I'm leaning towards executing the preemption in a subsequent cycle for unschedulable pods.

This seems to be aligned with what I wrote in my first comment in this thread.
Now my primary question is - can we ensure that the remaining pods will actually not be send to the back of the queue, and instead land on its front (obviously assuming there aren't higher priority ones added in the meantime)?

Author: > Now my primary question is - can we ensure that the remaining pods will actually not be send to the back of the queue, and instead land on its front (obviously assuming there aren't higher priority ones added in the meantime)?

It's a bit tricky. We could send the pod group directly to the activeQ after such attempt, but it would have a newer timestamp (we calculate the order based on the timestamp), so it will be considered later than the equal priority pod(group)s. We could use the previous timestamp, although this might be unintuitive in the code, it shouldn't cause any problems.

Reviewer 5: > although this might be unintuitive in the code, it shouldn't cause any problems.

That was primarily my question.
If this "should work", I think it's effectively what we should do to strengthen the story of "workload is scheduled together".

Author: Added to the algorithm description that the preemption will be attempted in the subsequent cycle and the pod group will be requeued with the old timestamp in case it was mostly scheduled.

Reviewer 5: Thank you - this looks great now!

       from the scheduler's perspective. Alternatively, always attempting preemption to free up space,
       even for schedulable groups, would be unnecessarily disruptive and delay the startup of the group. 
      
       Unschedulable pods will be requeued using their old timestamp, meaning the subsequent scheduling
       and preemption attempt should start immediately after the current cycle,
       unless a higher-priority pod or pod group comes in between.

     * If this is a subsequent scheduling attempt (i.e., some pod group members were already scheduled
       when the cycle started), but some pods were unschedulable, the scheduler will attempt
       [Workload-aware Preemption](#workload-aware-preemption).

       * If the preemption can accommodate more pods, it will be actuated (as described below)
         and all considered pods, including the schedulable ones, will be moved back to the scheduling queue.

       * If preemption cannot help, schedulable pods will be moved to binding,
         just as in the initial scheduling attempt case.
       
       This aligns with real applications, where attempting preemption to secure capacity for as many additional,

>        remaining pods as possible is preferable to binding only a schedulable subset of pods.

Reviewer 5: nit:

... if soon after that we will schedule another set unblocked by preemption.

Or something along those lines.

Author: Done

"all-or-nothing" (`minCount`) checks will be skipped; i.e., we will try to
schedule as many pods from such PodGroup as possible.

#### Workload-aware Preemption

> Proper preemption support is a critical requirement for moving Gang Scheduling to Beta.

Reviewer 7: I think it's worth to add a paragraph about per-pod PostFilter status. Currently we skip it only for preemption plugin, but execute it for other plugins.

They need to eventually migrate to the PodGroup level PostFilter (introduced in WAP KEP) as future WAS features will require repeating the scheduling process several times, so per-pod PostFilter will be eventually disabled for PodGroups.

It's possible that the migration will come with the PostFilter refactoring effort which may simplify that process.

Reviewer 5: This comment should go to WAS KEP, not to this KEP.

/cc @Argh4k

Reviewer 7: It's definitely a comment for WAP, but this KEP also mentions disabling per-pod default preemption, so it may be worth to mention that other PostFilter executions stays (agree it does not have to be a separate section).

Author: > It's definitely a comment for WAP, but this KEP also mentions disabling per-pod default preemption, so it may be worth to mention that other PostFilter executions stays (agree it does not have to be a separate section).

Will other PostFilter executions stay? Didn't want we to move all PostFilter use cases to PodGroupPostFilter?

Also, I don't think this KEP should focus that much on the WAP KEP outcomes. We can always adjust the wording here and there later during the release cycle.

Reviewer 5: > Also, I don't think this KEP should focus that much on the WAP KEP outcomes. We can always adjust the wording here and there later during the release cycle.

+1 - let's move all those things to WAP KEP

Therefore, we optimistically retry the Workload Scheduling Cycle if any member's condition improves.

It might be beneficial to retry the pod group without being triggered by any cluster event,
because a single Workload Scheduling Cycle cannot determine if the placement does not really exist,
especially for heterogeneous workloads or inter-pod dependencies.
To avoid introducing subtle errors in the initial implementation,
we can start by skipping the Queueing Hints mechanism and relying solely on the backoff time.
As a later step, we can decide whether to re-enable Queueing Hints for PodGroups with homogeneous pods only.

### Feature gates merge in v1.37

Initially, the KEP was introduced with two feature gates: `GenericWorkload` and `GangScheduling`.
We identified that such a split is artificial and enabling only `GenericWorkload`
might provide unclear results (as it enables the entire API but only a subset of kube-scheduler functionalities).

> In v1.37, both feature gates will be merged into one: `GenericWorkload`. Moreover, since workload-aware preemption

Reviewer 9: This is going to be sorta difficult to track. Is the goal that WAP will graduate at same time as gang scheduling?

Then I would maybe suggest we merge the KEPs also. The point of having separate feature gates for KEPs is to enable the features in isolation.

Author: > Is the goal that WAP will graduate at same time as gang scheduling?

Yes. The preemption mechanism is necessary and we don't see a point in enabling the feature that doesn't support it. That's why having separate feature gates doesn't make sense, as they would need to graduate and be enabled in exactly the same way.

> Then I would maybe suggest we merge the KEPs also.

This KEP is already kinda large and operates on a different abstraction layer than the WAP KEP. Personally, I would keep them separate, pointing to each other where needed, unless there is a strong reason to merge them.

Reviewer 5: +1 to Maciek
They are both pretty complex and merging them would result in a giant beast hard to read.

That being said, we should add a bit note at the beginning of the PRR section that these 2 KEPs will progress in lockstep.

Reviewer 9: > That being said, we should add a bit note at the beginning of the PRR section that these 2 KEPs will progress in lockstep.

+1.

Will we keep the WAP KEP up to date? Remove the feature gate and update design in tandem with this one.

Reviewer 8: Yes, I am working on updating WAP KEP for 1.37.

Author: Added a few more sentences here and there in the KEP regarding both KEPs graduations

- `PodGroup` status is updated correctly
- `PodGroup` is garbage collected when the replica is deleted
  
With Workload Scheduling Cycle features, we will significantly expand test coverage to verify:

- Pods referencing a `PodGroup` (both gang and basic policies) are correctly processed via the Workload Scheduling 
  Cycle.
- `PodGroup` queuing ensures that all available members are retrieved and processed correctly.
- Deadlocks and livelocks do not occur when multiple gangs compete for resources or interleave with standard pods.
- Failed pod groups are requeued correctly and retry successfully when resources become available.
- Scheduler correctly captures updated `minCount` value for pending pod groups, potentially unblocking them from PreEnqueue.

We will also benchmark the performance impact of these changes to measure:

- The scheduling throughput of the workload scheduling, including gang and basic policies.

> ##### e2e tests

Reviewer 4: Can you please make sure to update both integration and e2e sections with the tests you've already written? Those are required for beta graduation.

Author: Updated with the links and wrote what tests are remaining, but needed for beta

- `PodGroup` status is updated correctly
- `PodGroup` is garbage collected when the replica is deleted
  
With Workload Scheduling Cycle features, we will significantly expand test coverage to verify:

- Pods referencing a `PodGroup` (both gang and basic policies) are correctly processed via the Workload Scheduling 
  Cycle.
- `PodGroup` queuing ensures that all available members are retrieved and processed correctly.
- Deadlocks and livelocks do not occur when multiple gangs compete for resources or interleave with standard pods.
- Failed pod groups are requeued correctly and retry successfully when resources become available.
- Scheduler correctly captures updated `minCount` value for pending pod groups, potentially unblocking them from PreEnqueue.

> - Changes to `minCount` value during Workload Scheduling Cycle doesn't affect the current cycle, but the subsequent ones.

Reviewer 5: This sounds like something fairly tricky to test and inherently racy (I mean if this requirement wouldn't be satisfied, it will be very hard to make this test consistently fail).

Author: That's right. Removed it from this list

  by kube-scheduler
- Implement workload-aware preemption to avoid premature preemptions
  and deliver a proper algorithm of handling PodGroup preemptions.
- Implement PodGroup queueing algorithm.
- Both `Workload` and `PodGroup` APIs are integrated (alpha) with at least one true workload[^6] controller.
- There are no blockers or concerns from the true workload controller[^6] communities,
  proving that the Workload and PodGroup APIs serve as a solid baseline for their use cases.
- A deletion protection mechanism is implemented for `PodGroup` objects and finalizer is added to the API.
- All e2e tests for `PodGroup` are added.
- Performance tests are created and are being run in CI to protect against regressions.

#### GA

- All issues and gaps identified as feedback during beta are resolved
- Promote the e2e API tests to conformance together with a test for gang scheduling behavior.

This KEP and workload-aware preemption ([KEP-5710]) are tightly coupled and controlled by the same feature gate.
They graduate in lockstep, meaning the graduation criteria of [KEP-5710] should be also met when targeting to a release.

### Upgrade / Downgrade Strategy

> This KEP is completely additive and can safely fallback to the original behavior on downgrade.

Reviewer 4: Can you explain how the merge of the two FGs (GangScheduling and GenericWorkload) will affect upgrade/downgrade paths? We error out if you specify unknown FG, which might be a problem during upgrade, and similarly the lack of the other FG during downgrade might cause unexpected behaviors, right?

Reviewer 4: Given that you're entirely removing the v1alpha2 version, should we also call out here explicitly that users should remove all the resources stored with that version, b/c they won't be able to reach to them after the upgrade?

Reviewer 5: > Given that you're entirely removing the v1alpha2 version, should we also call out here explicitly that users should remove all the resources stored with that version, b/c they won't be able to reach to them after the upgrade?

That's actually a good question. v1alpha2 was added as part of Alpha.
We had multiple cases in the past when doing breaking changes between Alpha and Beta and removing Alpha APis together. I think we were generally not describing that in the PRR (or at least I wasn't bringing that up) according to the rule: "if you're using Alpha features, you do that on your own risk".

I'm not against mentioning things like this, but is that a requirement? Might worth discussing during PRR meeting.

Author: Added explanation of both cases

Reviewer 4: > I'm not against mentioning things like this, but is that a requirement? Might worth discussing during PRR meeting.

Normally, I'd agree with that "if you're using Alpha features, you do that on your own risk", but somehow I feel a lot of users is actually testing this functionality and that wording in this section hopefully serve as a good reminder for them.

Reviewer 4: On top of that, is the FG merge, which might be a surprise to some, and again I know the alpha risk comment applies 😉

Reviewer 5: OK - the "exposure to community" comment makes sense to me.
I think that many of those people don't read KEPs, but it won't hurt.

And Maciek already added it anyway :)

- Providing "optimal enough" placement by considering all pods from a gang together
- Avoiding livelock scenario when multiple workloads are being scheduled at the same time
  by kube-scheduler
- Implement workload-aware preemption to avoid premature preemptions
  and deliver a proper algorithm of handling PodGroup preemptions.
- Implement PodGroup queueing algorithm.
- Both `Workload` and `PodGroup` APIs are integrated (alpha) with at least one true workload[^6] controller.
- A deletion protection mechanism is implemented for `PodGroup` objects and finalizer is added to the API.
- All e2e tests for `PodGroup` are added.
- Performance tests are created and are being run in CI to protect against regressions.

#### GA

> - All issues and gaps identified as feedback during beta are resolved

Reviewer 5: Promotion tests to conformance:
1) API tests
2) I would also push for having some gang-scheduling test to be part of conformance.

Author: Added a sentence about both

###### How can this feature be enabled / disabled in a live cluster?

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: GenericWorkload

>   - Components depending on the feature gate:

Reviewer 2: We should add `kube-controller-manager` here - PodGroup protection controller (added as a part of KEP-5832 that got merged with this KEP) runs if and only if `GenericWorkload` flag is enabled.

Ditto in `kep.yaml`.

I guess it would be also good to mention this in the `Scalability` section for the sake of completeness.

Author: Added here, to `kep.yaml` and mentioned in `Scalability`

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. The `GenericWorkload` feature gate needs to be switched off to disable the API changes, admission and gang scheduling
functionality. However, the content of `spec.schedulingGroup` fields in Pod objects will not be
cleared, as well as the existing Workload and PodGroup objects will not be deleted.

###### What happens if we reenable the feature if it was previously rolled back?

The feature should start working again.

> However, the user need to remember that some Workload and PodGroup objects could already be stored

Reviewer 4: ```suggestion
However, the user needs to remember that some Workload and PodGroup objects could already be stored
```

in etcd and may affect the behavior of some of the existing workloads.

###### Are there any tests for feature enablement/disablement?

The feature has unit tests that verifies enablement and disablement of the `schedulingGroup` field in Pod API:
https://github.com/kubernetes/kubernetes/blob/cc9fd614c429f43e67fd8ded5e160fd784d860a8/pkg/registry/core/pod/strategy_test.go#L3904

> Note that gang-scheduling itself is purely in-memory feature, so features themselves are enough.

Reviewer 4: Maybe something like: 
```suggestion
Note that gang-scheduling itself is purely in-memory feature, so these feature tests are enough.
```
will be better sounding?

- `process_start_time_seconds`: Frequent resets of this metric indicate that the scheduler process is crashing and 
  restarting (crash loop).
- `scheduler_pod_scheduling_duration_seconds`: A significant regression in P99 latency for standard (non-gang) pods 
  would indicate that the overhead of the new logic is unacceptable.
- `scheduler_podgroup_schedule_attempts_total`: Consistently 
  high failure rates for valid gangs compared to successful attempts.
- `scheduler_pending_entities{type="podgroup"}`: Unexpectedly high value may indicate issues
  with the pod group queueing algorithm.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

We'll perform manual testing of the upgrade -> downgrade -> upgrade path using the following sequence:

1. Start a local Kubernetes v1.36 cluster with GenericWorkload feature gate disabled (default 
behavior).
2. Attempt to create a Pod with `spec.schedulingGroup` set.
3. The `spec.schedulingGroup` field is dropped by the API server. The pod is created successfully
   but without the `schedulingGroup` reference, resulting in immediate standard scheduling (one-by-one).

> 4. Restart/Upgrade API Server and Scheduler to v1.37 (with feature gates enabled).

Reviewer 5: I would remove brackets - the FG will be disabled by default in 1.37 because the API will be disabled by default.
So let's clarify this.

Author: Done

5. Create two PodGroup objects: `gang-test-A` and `gang-test-B` (both with `minCount=2`).
6. Create a Pod `test-pod-1` with `spec.schedulingGroup` pointing to `gang-test-A`.
7. The Pod stays in `Pending` state (waiting for the gang). Verify that
   `scheduler_pending_entities{type="podgroup", queue="gated"}` metric is incremented.
8. Create a Pod `test-pod-2` pointing to the same pod group.
9. Both pods are scheduled successfully in the same cycle (Gang Scheduling works). 

> 10. Downgrade API Server and Scheduler to v1.36 with feature gates disabled.

Reviewer 7: ```suggestion
10. Downgrade API Server and Scheduler to v1.36 with feature gates (`GenericWorkload` and `GangScheduling` defined in that release) disabled.
```

11. Create `test-pod-3` pointing to `gang-test-B`. Note: We use a pod group created in step 5 because creating new
    PodGroup objects is disabled.
12. The pod is scheduled immediately (PodGroup logic is ignored because the schedulingGroup field is dropped by
    the v1.36 API server). If Gang Scheduling were active, this pod would hang pending waiting for a second member.

> 13. Upgrade API Server and Scheduler back to v1.37 (feature gates enabled).

Reviewer 5: Again - remove brackets, because you need to explicitly enable FG.

Author: Done

We'll perform manual testing of the upgrade -> downgrade -> upgrade path using the following sequence:

1. Start a local Kubernetes v1.36 cluster with GenericWorkload feature gate disabled (default 
behavior).
2. Attempt to create a Pod with `spec.schedulingGroup` set.
3. The `spec.schedulingGroup` field is dropped by the API server. The pod is created successfully
   but without the `schedulingGroup` reference, resulting in immediate standard scheduling (one-by-one).
4. Restart/Upgrade API Server and Scheduler to v1.37 with feature gates enabled.
5. Create two PodGroup objects: `gang-test-A` and `gang-test-B` (both with `minCount=2`).
6. Create a Pod `test-pod-1` with `spec.schedulingGroup` pointing to `gang-test-A`.
7. The Pod stays in `Pending` state (waiting for the gang). Verify that
   `scheduler_pending_entities{type="podgroup", queue="gated"}` metric is incremented.
8. Create a Pod `test-pod-2` pointing to the same pod group.
9. Both pods are scheduled successfully in the same cycle (Gang Scheduling works). 
10. Downgrade API Server and Scheduler to v1.36 with feature gates
    (`GenericWorkload` and `GangScheduling` defined in that release) disabled.

> 11. Create `test-pod-3` pointing to `gang-test-B`. Note: We use a pod group created in step 5 because creating new

Reviewer 4: I don't think this will work for you. Remember that v1.37 will have either v1alpah3 or v1beta1 versions, both of which are not available in 1.36, so this might not work for you, or am I missing something?

Reviewer 5: Great catch.

The fix for that is changing point (10) above - instead of running 1.36 with FG disabled, let's simply run 1.37 with FG disabled.
That would address this problem.

Author: Adjusted the algorithm to use v1.37 only

Reviewer 4: Not quite upgrade, but it serves its purpose. We should be able to update this flow in the next iteration.

    PodGroup objects is disabled.
12. The pod is scheduled immediately (PodGroup logic is ignored because the schedulingGroup field is dropped by

>     the v1.36 API server). If Gang Scheduling were active, this pod would hang pending waiting for a second member.

Reviewer 4: No, it would not, b/c the podgroup would not be visible due to version mismatch. See my previous comment.
